File: 2020/8.jl
include("utils.jl")
using .AoCUtils: readInput

function loadInput()
    filename = "prova.txt"
    filename = "input_8.txt"
    lines = readInput(filename)
    program = Vector{Vector{Union{String, SubString, Int}}}()
    for l in lines
        l = split(l)
        push!(program, [l[1], parse(Int, l[2])])
    end
    return program
end

function run(program::Vector{Vector{Union{String, SubString, Int}}})
    executed_lines = Int[]
    acc = 0
    idx = 1
    inf_loop = false
    while idx <= length(program)
        if idx in executed_lines
            inf_loop = true
            break
        else
            push!(executed_lines, idx)
        end

        line = program[idx]
        if line[1] == "nop"
            idx += 1
        elseif line[1] == "acc"
            acc += line[2]
            idx += 1
        elseif line[1] == "jmp"
            idx += line[2]
        end
    end
    return acc, inf_loop
end

function part1(program::Vector{Vector{Union{String, SubString, Int}}})
    acc, inf_loop = run(program)
    println("🎄 Part 1: $(acc)")
end

function modify_line(line::SubString{String})
    if line == "nop"
        return "jmp"
    else
        return "nop"
    end
end

function part2(program::Vector{Vector{Union{String, SubString, Int}}})
    idxs = findall(x->x[1] == "nop" || x[1] == "jmp", program)
    new_program = deepcopy(program)
    for idx in idxs
        new_program[idx][1] = modify_line(new_program[idx][1])
        acc, loop = run(new_program)
        if !loop
            println("🎄🎅 Part 2: $(acc)")
            break
        end
        new_program[idx][1] = modify_line(new_program[idx][1])
    end     
end

println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
println("⛄        Day 8         ⛄")
println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

inputs = loadInput()
@time part1(inputs)
@time part2(inputs)

File: 2020/8.py
import time
from utils import readInput

class Compiler:
    def __init__(self):
        self.code = {}
        self.init()
        with open("input_8.txt") as f:
            for i, l in enumerate(f):
                items = l.split(" ")
                self.code[i] = (items[0], int(items[1]))

    def init(self):
        self.code_run = []
        self.accumulator = 0
        self.pointer = 0

    def list(self):
        for v in self.code.values():
            print (v[0], v[1])

    def run(self):
        while self.pointer < len(self.code):
            if not self.inst_run():
                return False
            self.interpret()
        return True

    def interpret(self):
        inst, param = self.code[self.pointer]
        if inst == "nop":
            self.pointer += 1
        elif inst == "acc":
            self.pointer += 1
            self.accumulator += param
        elif inst == "jmp":
            self.pointer += param
        else:
            raise(RuntimeError("Command not found {} {}".format(inst, param)))

    def inst_run(self):
        if self.pointer not in self.code_run:
            self.code_run.append(self.pointer)
            return True
        else:
            return False

    def fix_code(self):
        orig_code = dict(self.code)
        for i in orig_code:
            self.init()
            self.code = dict(orig_code)
            inst, param = self.code[i]
            if inst == "nop" and param != 0:
                self.code[i] = ("jmp", param)
            elif inst == "jmp":
                self.code[i] = ("nop", param)
            else:
                continue
            res = self.run()
            if res:
                return self.accumulator

def part1():
    c = Compiler()
    c.run()
    print ("🎄 Part 1: {}".format(c.accumulator))

def part2():
    c = Compiler()
    print ("🎄🎅 Part 2: {}".format(c.fix_code()))

print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
print("⛄        Day 8         ⛄")
print("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

t0 = time.time()
part1()
print ("Time: {:.5f}".format(time.time()-t0))

t0 = time.time()
part2()
print ("Time: {:.5f}".format(time.time()-t0))
